service/history/api/multioperation/api.go

```diff
@@ -108,64 +112,116 @@ func Invoke(
-		updateReq,
-	)
-	startResp, err := starter.Invoke(
 		ctx,
-		func(lease api.WorkflowLease) error {
-			workflowCtx = lease.GetContext()
-			_, updateErr = updater.ApplyRequest(ctx, lease)
-			return updateErr
+		currentWorkflowLease,
+		func(lease api.WorkflowLease) (*api.UpdateWorkflowAction, error) {
+			return updater.ApplyRequest(ctx, lease)
 		},
+		nil,
 	)
 	if err != nil {
-		// TODO: send per-operation error details
-		return nil, fmt.Errorf("failed to start workflow: %w", err)
+		return onUpdateError(err, updater, startOpResp)
+	}
+
+	// wait for the update to complete
+	updateResp, err := updater.OnSuccess(ctx)
+	if err != nil {
+		// TODO: send Update outcome failure instead
+		return nil, fmt.Errorf("failed to complete Workflow Update: %w", err)
+	}
+
+	return &historyservice.ExecuteMultiOperationResponse{
+		Responses: []*historyservice.ExecuteMultiOperationResponse_Response{
+			startOpResp,
+			{
+				Response: &historyservice.ExecuteMultiOperationResponse_Response_UpdateWorkflow{
+					UpdateWorkflow: updateResp,
+				},
+			},
+		},
+	}, nil
+}
+
+func startAndUpdateWorkflow(
+	ctx context.Context,
+	shardContext shard.Context,
+	workflowCache wcache.Cache,
+	starter *startworkflow.Starter,
+	updater *updateworkflow.Updater,
+) (*historyservice.ExecuteMultiOperationResponse, error) {
+	var workflowKey definition.WorkflowKey
+	var workflowCtx workflow.Context
+	var namespaceID namespace.ID
+	var updateErr error
+
+	// hook to apply workflow update before the workflow is being persisted
+	hook := func(lease api.WorkflowLease) error {
+		workflowCtx = lease.GetContext()
+		workflowKey = workflowCtx.GetWorkflowKey()
+		namespaceID = lease.GetMutableState().GetNamespaceEntry().ID()
+		_, updateErr = updater.ApplyRequest(ctx, lease) // ignoring UpdateWorkflowAction return since Start will create WFT
+		return updateErr
 	}
+
+	// start workflow and apply update via `hook`
+	startResp, err := starter.Invoke(ctx, hook)
 	startOpResp := &historyservice.ExecuteMultiOperationResponse_Response{
 		Response: &historyservice.ExecuteMultiOperationResponse_Response_StartWorkflow{
 			StartWorkflow: startResp,
 		},
 	}
+	if err != nil {
+		// an update error occurred
+		if updateErr != nil {
+			return onUpdateError(updateErr, updater, startOpResp)
+		}
 
-	if updateErr != nil {
-		updateResp := updater.OnError(updateErr)
-		return &historyservice.ExecuteMultiOperationResponse{
-			Responses: []*historyservice.ExecuteMultiOperationResponse_Response{
-				startOpResp,
-				{
-					Response: &historyservice.ExecuteMultiOperationResponse_Response_UpdateWorkflow{
-						UpdateWorkflow: updateResp,
-					},
-				},
-			},
-		}, nil
+		// a start error occurred
+		// TODO: send per-operation error details
+		return nil, fmt.Errorf("failed to start workflow: %w", err)
 	}
 
-	// Without this, there's no Update registry on the call from Matching back to History.
+	// without this, there's no Update registry on the call from Matching back to History
```

Comment:
I've noticed quite a few places in the history service codebase where comments that were not written as sentences became ambiguous, because multiple such comments were placed on consecutive lines and it wasn't clear what was a  sentence break vs. merely a line break. So I actually think that we should write comments as sentences to avoid ambiguities. There may be some evidence in support of this in https://github.com/temporalio/temporal/commit/cb2139cefb6cda5e1a0a44dae70a0c644f8d6459.

Reply:
I've noticed that problem, too.

I _usually_ write lower-case/no period statements when it's a single line and complete sentences when it's more than one line. But then one can morph into the other over time and cause issues. So I agree that writing sentences from the start makes it less error-prone. 

At the same time, I like the ease/low friction of not having to adhere to a perfect sentence syntax. It's also very habitual (for me at least).

Reply:
In this particular case here, my IDE hides the fact from me that these are two statements through syntax highlighting
<img width="677" alt="Screenshot 2024-04-22 at 10 35 17 AM" src="https://github.com/temporalio/temporal/assets/159852/8747d347-ef35-4e32-b2a0-8f3750453669">

Reply:
Let's have a conversation outside of this PR.